### ISSUE TITLE
Passing one interceptor only per request using okhttp Call proxies

### ISSUE THREAD

**Issue by vpasquier** [square/retrofit#1422]
Hi,

I've seen here #261 the work and discussions about how to add header per request and not globally for all requests. But the solution of adding @header parameters is not convenient for the purpose of my use case (I have lots of endpoints method to implement and headers as well).

Is there a way to pass to the okhttp Call proxies an interceptor (for one request of this proxy only) dynamically?

Thank you!

**Comment by JakeWharton** [comment]
You want to pass an interceptor as a parameter?

**Comment by vpasquier** [comment]
Nop that is what I want to avoid.

I would like to add interceptor like we do for networkInterceptor or interceptor but not as global for all requests, just only per request. And the thing is that apparently there is no mean to pass this "local" sporadic interceptor with the Call proxy.

Except using the object com.squareup.okhttp.Request. But I would like to use Retrofit.

**Comment by JakeWharton** [comment]
How do you associate the interceptor with the desired endpoint for which it is to apply?

**Comment by vpasquier** [comment]
Oh yeah I understand your point. It's true that on my proxy I'm going to have multiple endpoint and it could be just applied on this group of endpoints. (why not..)

So in another way, in order to control all the global interceptors to set them just once per request using Retrofit proxies, what would be the best solution to do this? I wanted to manage my interceptors, but there is no id for them as they are in a list (com.squareup.okhttp.OkHttpClient#interceptors or network ones).

What could be the best? Setting global interceptors is interesting but sometimes you would like to set just per request a header.

**Comment by JakeWharton** [comment]
I still don't understand the API you are looking for. You don't want a global interceptor but you also don't want a per-call interceptor. Can you phrase the request in the form of code?

**Comment by vpasquier** [comment]
Ok sorry.

I want a per-call interceptor. How can I achieve that ? Is the only solution is to use the @header parameters?

**Comment by JakeWharton** [comment]
You can use a `@Header` param for a dynamic, per-endpoint value, the `@Headers` method annotation for a fixed, per-endpoint value, or use an global OkHttp interceptor and match on the request path (or anything else) for modifying the request in any arbitrary way. We don't have any other ways to customize a single endpoint.

**Comment by vpasquier** [comment]
First of all thank you for the client and this release beta3. Huge work!

Unfortunately I still stuck with Retrofit proxies to access to the original request, update the headers per call and to retrieve easily the original response.

To explain more, because it's globally the same subject: **how to get/update original request and response with proxies.**

### Request

As explained here #1367, we need to properly access to the original request but also to be able to update it on the fly per call. ([reflection is mandatory](https://github.com/nuxeo/nuxeo-java-client/blob/master/nuxeo-java-client/src/main/java/org/nuxeo/java/client/api/objects/NuxeoEntity.java#L151-L156) for now)

#####Request Headers

The problem of the 'dynamic' `@header` parameters are that the maintenance of such apis is impossible. We have to duplicate all our methods in order to insert all possible headers and it doesn't even allow to add custom header to apis in general (we should have a header with explicit name inside the annotation). 

The principle of `@header` can be nice for certain use case but when you have multiple endpoints like [this](https://github.com/nuxeo/nuxeo-java-client/blob/master/nuxeo-java-client/src/main/java/org/nuxeo/java/client/api/methods/RepositoryAPI.java#L38-L146), you cannot maintain all of those headers you would like to put per request.

-> The need of having the possibility to change the request within the proxies is necessary to use Retrofit proxies in general with lots of APIs.

In general, could it be possible to provide to the `retrofit2.RequestFactory` some informations like header through `retrofit2.Call.execute()` method ? And not just taking into account the annotations in the proxied interface?

### Response

As mentioned here #1385, following your comment around the `Call.Factory`, I couldn't find a way to do it properly.

*I really hope we can find a solution, I really love the way Retrofit is working. I could work on it if you haven't time and will do the best as I can and of course, if I misunderstood something about the Retrofit proxies, don't hesitate!:)*

**Comment by vpasquier** [comment]
Hello Jake,

Just to know if it would be possible. Something like this (just a awful draft) https://github.com/vpasquier/retrofit/commit/ab75508136b9f90b5e952429acf79d1b36096264

I could work on that but tell me its interesting for you as well to have dynamic headers.

**Comment by JakeWharton** [comment]
Call should not appear to be mutable.

**Comment by vpasquier** [comment]
I understand. But do you have an idea how to achieve it? Do you have hint to achieve it? I can find an idea and work on it but I wanted to know what you think about my previous message and this feature in the principle. The call is mutable but it has been filled by annotations by the parser.

**Comment by andretietz** [comment]
I do actually have an idea how to achieve this and I want to request this as well:

Problem is, that within the retrofit2.CallAdapter there's no option to modify/intercept the okhttp3.Request building. To solve @vpasquier s (and my problem), we need to at least be able to add a Tag to the request object.   

Problem I face at the moment is, that when you implement the CallAdapter<T> interface, you have to impement the adapt method, which returns an already set Call<T> Object (which is created by retrofit as OkHttpCall) and contains the created and not accessible okhttp3.Request object already.

Some possible options that could handle this issue (to satify both your points):
* Instead of implementing retrofit2.CallAdapter.adapt(retrofit2.Call<T> call) having to implement something like this: retrofit2.Call<T> adapt(okhttp3.Request request) so we're able to modify it before it gets created. 
* having an optional method (so CallAdapter could be an abstract i.e.) okhttp3.Request modifyRequest(okhttp3.Request(Builder?!) rawRequest)
* set a tag to the Call<T> which will be handed into the request when the request is created

There are lots of possibilities, this were just a few. 

Then this modified request can be reached in the okhttp3.Interceptor then. (to full fill @vpasquier s requirement) and it's immutable (to full fill @JakeWharton s)

What do you think?

**Comment by JakeWharton** [comment]
A CallAdapter exists solely to change the return type of a service method. It runs before any `Request` is even created and does so synchronously when the service method is called. The `Request` is lazy and does not get created until you call `execute` or `enqueue` (or the adapted type, like RxJava's `Observable` does it for you).

I'm open to a hook into the `Request` creation but it likely needs a good bit of thought as it will have to be something new.

**Comment by andretietz** [comment]
Yeah, I've taken a look to the code on the weekend and I agree to that. It's not as easy as I thought it would. But still possible. If I find some time, I'll try to create a solution

**Comment by andretietz** [comment]
This is one step in the direction already.
https://github.com/square/retrofit/pull/1484

What could work now:

in the adapt method we could take the request (using the new retrofit2.Call#request method), modify it as we want (using okhttp.Request#newBuilder) and pass it into an own implementation of retrofit2.Call which we can then return in the CallAdapter#adapt method. So in general this is working already, but it's not the best solution since everyone has to implement it's own "special" retrofit.Call.
